src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -195,76 +195,96 @@ OMapInnerNode::rm_key(omap_context_t oc, const std::string &key)
-  const std::optional<std::string> &start,
+  const std::optional<std::string> &first,
+  const std::optional<std::string> &last,
   omap_list_config_t config)
 {
   LOG_PREFIX(OMapInnerNode::list);
-  if (start) {
-    DEBUGT("start={}, this: {}", oc.t, *start, *this);
+  if (first && last) {
+    DEBUGT("first: {}, last: {}, this: {}", oc.t, *first, *last, *this);
+    assert(*first <= *last);
+  } else if (first) {
+    DEBUGT("first: {}, this: {}", oc.t, *first, *this);
+  } else if (last) {
+    DEBUGT("last: {}, this: {}", oc.t, *last, *this);
   } else {
     DEBUGT("this: {}", oc.t, *this);
   }
 
-  auto child_iter = start ?
-    get_containing_child(*start) :
+  auto first_iter = first ?
+    get_containing_child(*first) :
     iter_cbegin();
-  assert(child_iter != iter_cend());
+  auto last_iter = last ?
+    get_containing_child(*last) + 1:
+    iter_cend();
+  assert(first_iter != iter_cend());
 
   return seastar::do_with(
-    child_iter,
-    iter_cend(),
+    first_iter,
+    last_iter,
+    iter_t(first_iter),
     list_bare_ret(false, {}),
-    true,
-    start,
+    first,
+    last,
     [this, oc, config](
-      auto &biter,
-      auto &eiter,
+      auto &fiter,
+      auto &liter,
+      auto &iter,
       auto &ret,
-      bool &first_entry,
-      auto &start)
+      auto &first,
+      auto &last)
     {
       auto &complete = std::get<0>(ret);
       auto &result = std::get<1>(ret);
       return trans_intr::repeat(
-	[&, config, oc, this]() -> list_iertr::future<seastar::stop_iteration> {
-	  if (biter == eiter  || result.size() == config.max_result_size) {
-	    complete = biter == eiter;
-	    return list_iertr::make_ready_future<seastar::stop_iteration>(
-	      seastar::stop_iteration::yes);
-	  }
-	  auto laddr = biter->get_val();
-	  return omap_load_extent(
-	    oc, laddr,
-	    get_meta().depth - 1
-	  ).si_then([&, config, oc] (auto &&extent) {
-	    return extent->list(
-	      oc,
-	      first_entry ? start : std::nullopt,
-	      config.with_reduced_max(result.size())
-	    ).si_then([&, config](auto &&child_ret) mutable {
-	      boost::ignore_unused(config);   // avoid clang warning;
-	      auto &[child_complete, child_result] = child_ret;
-	      if (result.size() && child_result.size()) {
-		assert(child_result.begin()->first > result.rbegin()->first);
+        [&, config, oc, this]() -> list_iertr::future<seastar::stop_iteration>
+      {
+        if (iter == liter || result.size() == config.max_result_size) {
+          complete = iter == liter;
+          return list_iertr::make_ready_future<seastar::stop_iteration>(
+            seastar::stop_iteration::yes);
+        }
+        auto laddr = iter->get_val();
+        return omap_load_extent(
+          oc, laddr,
+          get_meta().depth - 1
+        ).si_then([&, config, oc](auto &&extent) {
+          return extent->list(
+            oc,
+            iter == fiter ? first : std::nullopt,
+	    iter == liter - 1 ? last : std::nullopt,
```

Comment:
Looks incorrect to pass the `std::nullopt`s directly.

Reply:
addressed.